diff --git a/src/modules/module-echo-cancel.c b/src/modules/module-echo-cancel.c
index 7b33c47..def6476 100644
--- a/src/modules/module-echo-cancel.c
+++ b/src/modules/module-echo-cancel.c
@@ -1243,7 +1243,7 @@ int pipewire__module_init(struct pw_impl_module *module, const char *args)
 		goto error;
 	}
 
-	impl->monitor_mode = false;
+	impl->monitor_mode = true;
 	if ((str = pw_properties_get(props, "monitor.mode")) != NULL)
 		impl->monitor_mode = pw_properties_parse_bool(str);
 
