From 5b7bd93f0493e88637ff4a6ef1140a13ded05cc8 Mon Sep 17 00:00:00 2001
From: flightlessmango <flightlessmangoyt@gmail.com>
Date: Mon, 17 Feb 2025 21:18:04 +0100
Subject: [PATCH] Params: preset 4: disable some more options for steam deck

---
 src/overlay_params.cpp | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/overlay_params.cpp b/src/overlay_params.cpp
index 9fbe4a4fc4..cf4446a87f 100644
--- a/src/overlay_params.cpp
+++ b/src/overlay_params.cpp
@@ -1165,8 +1165,13 @@ void presets(int preset, struct overlay_params *params, bool inherit) {
          add_to_options(params, "frame_timing_detailed", "1");
          add_to_options(params, "network", "1");
          add_to_options(params, "present_mode", "0");
-         if ( deviceID == 0x1435 || deviceID == 0x163f )
+         // Disable some options if steamdeck
+         if ( deviceID == 0x1435 || deviceID == 0x163f ) {
             add_to_options(params, "gpu_fan", "0");
+            add_to_options(params, "gpu_junction_temp", "0");
+            add_to_options(params, "gpu_voltage", "0");
+            add_to_options(params, "gpu_mem_temp", "0");
+         }
 
          break;
 
