From 6dcd1b5c4bbd9fa1ff56337bff6d307568afc4d5 Mon Sep 17 00:00:00 2001
From: flightlessmango <flightlessmangoyt@gmail.com>
Date: Mon, 17 Feb 2025 20:28:18 +0100
Subject: [PATCH] amdgpu: current_gfxclk was not being collected in APUs

---
 src/amdgpu.cpp | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/src/amdgpu.cpp b/src/amdgpu.cpp
index ebd2667e1f..b3cc5b9375 100644
--- a/src/amdgpu.cpp
+++ b/src/amdgpu.cpp
@@ -133,6 +133,17 @@ void AMDGPU::get_instant_metrics(struct amdgpu_common_metrics *metrics) {
 			metrics->apu_cpu_temp_c = 0;
 		}
 
+		if( IS_VALID_METRIC(amdgpu_metrics->current_gfxclk) ) {
+			// prefered method
+			metrics->current_gfxclk_mhz = amdgpu_metrics->current_gfxclk;
+		} else if( IS_VALID_METRIC(amdgpu_metrics->average_gfxclk_frequency) ) {
+			// fallback 1
+			metrics->current_gfxclk_mhz = amdgpu_metrics->average_gfxclk_frequency;
+		} else {
+			// giving up
+			metrics->current_gfxclk_mhz = 0;
+		}
+
 		indep_throttle_status = amdgpu_metrics->indep_throttle_status;
 	}
 
