From f60145c288610145b172af4f0dff49697337abc4 Mon Sep 17 00:00:00 2001
From: flightlessmango <flightlessmangoyt@gmail.com>
Date: Mon, 17 Feb 2025 21:07:20 +0100
Subject: [PATCH] amdgpu: current_uclk was not being collected in APUs

---
 src/amdgpu.cpp | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/src/amdgpu.cpp b/src/amdgpu.cpp
index b3cc5b9375..4a2542298a 100644
--- a/src/amdgpu.cpp
+++ b/src/amdgpu.cpp
@@ -143,6 +143,17 @@ void AMDGPU::get_instant_metrics(struct amdgpu_common_metrics *metrics) {
 			// giving up
 			metrics->current_gfxclk_mhz = 0;
 		}
+		
+		if( IS_VALID_METRIC(amdgpu_metrics->current_uclk) ) {
+			// prefered method
+			metrics->current_uclk_mhz = amdgpu_metrics->current_uclk;
+		} else if( IS_VALID_METRIC(amdgpu_metrics->average_uclk_frequency) ) {
+			// fallback 1
+			metrics->current_uclk_mhz = amdgpu_metrics->average_uclk_frequency;
+		} else {
+			// giving up
+			metrics->current_uclk_mhz = 0;
+		}
 
 		indep_throttle_status = amdgpu_metrics->indep_throttle_status;
 	}
