diff -up udisks-2.9.4/src/tests/dbus-tests/test_job.py.bak udisks-2.9.4/src/tests/dbus-tests/test_job.py
--- udisks-2.9.4/src/tests/dbus-tests/test_job.py.bak	2021-09-29 18:00:31.000000000 +0200
+++ udisks-2.9.4/src/tests/dbus-tests/test_job.py	2023-06-02 15:14:50.691620552 +0200
@@ -51,6 +51,7 @@ class UdisksJobTest(udiskstestcase.Udisk
 
             time.sleep(0.1)
 
+    @udiskstestcase.tag_test(udiskstestcase.TestTags.UNSTABLE)
     def test_job(self):
         '''Test basic Job functionality and properties'''
 
