diff -Naur a/pyanaconda/display.py b/pyanaconda/display.py
--- a/pyanaconda/display.py
+++ b/pyanaconda/display.py
@@ -263,6 +263,13 @@
     :param str runres: a resolution specification string
     """
     try:
+        # Get product name
+        with open("/sys/devices/virtual/dmi/id/product_name") as f:
+            dmi = f.read().strip()
+
+        if "Jupiter" in dmi or "Galileo" in dmi or "NEXT" in dmi:
+            runres = "1280x800"
+
         log.info("Setting the screen resolution to: %s.", runres)
         mutter_display = MutterDisplay()
         mutter_display.set_resolution(runres)

diff -Naur a/pyanaconda/ui/gui/__init__.py b/pyanaconda/ui/gui/__init__.py
--- a/pyanaconda/ui/gui/__init__.py
+++ b/pyanaconda/ui/gui/__init__.py
@@ -584,6 +584,13 @@
         if not primary_monitor:
             return

+        with open("/sys/devices/virtual/dmi/id/product_name") as f:
+            dmi = f.read().strip()
+
+        if "Jupiter" in dmi or "Galileo" in dmi or "NEXT" in dmi:
+            util.setenv("GDK_SCALE", "1")
+            return
+
         monitor_geometry = primary_monitor.get_geometry()
         monitor_scale = primary_monitor.get_scale_factor()
         monitor_width_mm = primary_monitor.get_width_mm()
