--- a/pyanaconda/core/product.py
+++ b/pyanaconda/core/product.py
@@ -93,6 +93,14 @@ def get_product_values():
     product_name = config.get("Main", "Product")
     product_version = trim_product_version_for_ui(config.get("Main", "Version"))
 
+    try:
+        with open("/etc/bazzite-version", "r") as f:
+            bazzite_version = f.read().strip()
+            if bazzite_version:
+                product_version = bazzite_version
+    except Exception:
+        product_version = "44"
+
     # for use in device names, eg: "fedora", "rhel"
     product_short_name = shorten_product_name(product_name)
 
