diff --git a/src/util/00-mesa-defaults.conf b/src/util/00-mesa-defaults.conf
index 948c1ef..abe62c8 100644
--- a/src/util/00-mesa-defaults.conf
+++ b/src/util/00-mesa-defaults.conf
@@ -166,6 +166,12 @@ TODO: document the other workarounds.
             <option name="disable_uniform_array_resize" value="true" />
         </application>
 
+        <application name="The Chronicles of Riddick: Assault on Dark Athena" executable="DarkAthena.exe">
+           <option name="disable_uniform_array_resize" value="true" />
+            <option name="alias_shader_extension" value="GL_ATI_shader_texture_lod:GL_ARB_shader_texture_lod" />
+            <option name="allow_vertex_texture_bias" value="true" />
+        </application>
+
         <application name="Dying Light" executable="DyingLightGame">
             <option name="allow_glsl_builtin_variable_redeclaration" value="true" />
         </application>
@@ -364,6 +370,11 @@ TODO: document the other workarounds.
             <option name="lower_depth_range_rate" value="0.8" />
         </application>
 
+        <application name="The Spirit and The Mouse" executable="TheSpiritAndTheMouse.exe">
+            <!-- https://gitlab.freedesktop.org/mesa/mesa/-/issues/9738 -->
+            <option name="lower_depth_range_rate" value="0.8" />
+        </application>
+
         <application name="Penumbra: Overture" executable="Penumbra.exe">
             <!-- https://gitlab.freedesktop.org/mesa/mesa/-/issues/8425 -->
             <option name="allow_multisampled_copyteximage" value="true" />
@@ -718,6 +729,7 @@ TODO: document the other workarounds.
         <application name="gnome-shell" executable="gnome-shell">
             <option name="adaptive_sync" value="false" />
             <option name="v3d_nonmsaa_texture_size_limit" value="true" />
+            <option name="mesa_glthread" value="false"/>
         </application>
         <application name="Desktop — Plasma" executable="plasmashell">
             <option name="adaptive_sync" value="false" />
@@ -923,6 +935,39 @@ TODO: document the other workarounds.
             <option name="vk_x11_override_min_image_count" value="3" />
         </application>
 
+        <!-- Disable fp16 support for browsers, since there is too much
+             broken WebGL out there that uses the wrong precision.
+             Bonus workaround for Firefox bug #1845309. -->
+        <application name="Firefox" executable="firefox">
+            <option name="no_fp16" value="true" />
+        </application>
+        <application name="Firefox ESR" executable="firefox-esr">
+            <option name="no_fp16" value="true" />
+        </application>
+        <application name="Chromium" executable="chromium">
+            <option name="no_fp16" value="true" />
+        </application>
+        <application name="Google Chrome" executable="chrome">
+            <option name="no_fp16" value="true" />
+        </application>
+        <application name="Iceweasel" executable="iceweasel">
+            <option name="no_fp16" value="true" />
+        </application>
+        <application name="Epiphany" executable="epiphany">
+            <option name="no_fp16" value="true" />
+        </application>
+        <application name="Konqueror" executable="konqueror">
+            <option name="no_fp16" value="true" />
+        </application>
+        <application name="Falkon" executable="falkon">
+            <option name="no_fp16" value="true" />
+        </application>
+        <application name="Seamonkey" executable="seamonkey">
+            <option name="no_fp16" value="true" />
+        </application>
+        <application name="Waterfox" executable="waterfox">
+            <option name="no_fp16" value="true" />
+        </application>
     </device>
     <!-- vmwgfx doesn't like full buffer swaps and can't sync to vertical retraces.-->
     <device driver="vmwgfx">
@@ -1104,6 +1149,25 @@ TODO: document the other workarounds.
         <application name="Cyberpunk 2077" executable="Cyberpunk2077.exe">
             <option name="force_vk_vendor" value="-1" />
         </application>
+        <application name="Cyberpunk 2077" executable="Cyberpunk2077.exe">
+            <!--
+            Cyberpunk 2077 uses a watchdog thread to terminate
+            the process in case the render thread hasn't responded within 2 minutes.
+            This option speeds up shader compilation.
+            See: https://gitlab.freedesktop.org/mesa/mesa/-/issues/9241
+            -->
+            <option name="shader_spilling_rate" value="15" />
+        </application>
+        <application name="Elden Ring" executable="eldenring.exe">
+            <option name="fake_sparse" value="true" />
+        </application>
+        <!-- Needed to avoid XeSS code paths. -->
+        <application name="Marvel's Spider-Man Remastered" executable="Spider-Man.exe">
+            <option name="force_vk_vendor" value="-1" />
+        </application>
+        <application name="Hogwarts Legacy" executable="HogwartsLegacy.exe">
+            <option name="force_vk_vendor" value="-1" />
+        </application>
         <!--
         Disable 16-bit feature on zink and angle so that GLES mediump doesn't
         lower to our inefficent 16-bit shader support.  No need to do so for
diff --git a/src/util/00-radv-defaults.conf b/src/util/00-radv-defaults.conf
index 9434dba..3c3ae56 100644
--- a/src/util/00-radv-defaults.conf
+++ b/src/util/00-radv-defaults.conf
@@ -186,5 +186,9 @@ Application bugs worked around in this file:
         <application name="Rocket League" executable="RocketLeague">
             <option name="radv_zero_vram" value="true" />
         </application>
+
+        <application name="Forza Horizon 5" application_name_match="ForzaHorizon5.exe">
+            <option name="vk_x11_override_min_image_count" value="4" />
+        </application>
     </device>
 </driconf>
