diff -ur usr/share/jupiter-fan-control/fancontrol.py usr/share/jupiter-fan-control/fancontrol.py
--- usr/share/jupiter-fan-control/fancontrol.py	2022-12-23 13:14:48.861887067 -0800
+++ usr/share/jupiter-fan-control/fancontrol.py	2022-12-23 00:42:03.223412367 -0800
@@ -1,4 +1,4 @@
-#!/usr/bin/python -u
+#!/usr/bin/python3 -u
 """jupiter-fan-controller"""
 import signal
 import os
diff -ur usr/share/jupiter-fan-control/PID.py usr/share/jupiter-fan-control/PID.py
--- usr/share/jupiter-fan-control/PID.py	2022-12-23 13:14:48.861887067 -0800
+++ usr/share/jupiter-fan-control/PID.py	2022-12-23 00:42:03.222412347 -0800
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/python3
 #
 # This file is part of IvPID.
 # Copyright (C) 2015 Ivmech Mechatronics Ltd. <bilgi@ivmech.com>
@@ -64,7 +64,7 @@
         """Calculates PID value for given reference feedback
 
         .. math::
-            u(t) = K_p e(t) + K_i \int_{0}^{t} e(t)dt + K_d {de}/{dt}
+            u(t) = K_p e(t) + K_i \\int_{0}^{t} e(t)dt + K_d {de}/{dt}
 
         .. figure:: images/pid_1.png
            :align:   center
