diff --git a/etc/systemd/system/steamos-automount@.service b/etc/systemd/system/steamos-automount@.service
index dd700da..152e231 100644
--- a/etc/systemd/system/steamos-automount@.service
+++ b/etc/systemd/system/steamos-automount@.service
@@ -14,6 +14,7 @@ Description=Mount removable drive %i and notify Steam
 [Service]
 Type=oneshot
 RemainAfterExit=true
+EnvironmentFile=/etc/default/steamos-btrfs
 ExecStart=/usr/libexec/steamos-automount add %i
 ExecStop=/usr/libexec/steamos-automount remove %i
 ExecReload=/usr/libexec/steamos-automount retrigger %i
diff --git a/usr/lib/hwsupport/format-device.sh b/usr/lib/hwsupport/format-device.sh
index b136413..d3cbb16 100755
--- a/usr/lib/hwsupport/format-device.sh
+++ b/usr/lib/hwsupport/format-device.sh
@@ -2,6 +2,8 @@
 
 set -e
 
+source /etc/default/steamos-btrfs
+
 # If the script is not run from a tty then send a copy of stdout and
 # stderr to the journal. In this case stderr is also redirected to stdout.
 if ! tty -s; then
