diff --git a/usr/share/steamos/steamos-cursor-config b/usr/share/steamos/steamos-cursor-config
index 5891cd8..620f7d8 100644
--- a/usr/share/steamos/steamos-cursor-config
+++ b/usr/share/steamos/steamos-cursor-config
@@ -1,4 +1,4 @@
-32 5 3 steamos-cursor-32.png 100
-64 10 6 steamos-cursor-64.png 100
-128 20 12 steamos-cursor-128.png 100
-256 40 24 steamos-cursor-256.png 100
\ No newline at end of file
+32 5 3 usr/share/steamos/steamos-cursor-32.png 100
+64 10 6 usr/share/steamos/steamos-cursor-64.png 100
+128 20 12 usr/share/steamos/steamos-cursor-128.png 100
+256 40 24 usr/share/steamos/steamos-cursor-256.png 100
\ No newline at end of file
