diff --git a/usr/bin/amd_system_info b/usr/bin/amd_system_info
index ada6104..e40e82a 100755
--- a/usr/bin/amd_system_info
+++ b/usr/bin/amd_system_info
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 # script from AMD to use when reporting bugs to them
 
diff --git a/usr/bin/jupiter-biosupdate b/usr/bin/jupiter-biosupdate
index c0665c4..a8d6059 100755
--- a/usr/bin/jupiter-biosupdate
+++ b/usr/bin/jupiter-biosupdate
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 set -e
 
 ## Environment overrides
diff --git a/usr/bin/jupiter-check-support b/usr/bin/jupiter-check-support
index bd56a57..4571cc9 100755
--- a/usr/bin/jupiter-check-support
+++ b/usr/bin/jupiter-check-support
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -e
 
diff --git a/usr/bin/jupiter-controller-update b/usr/bin/jupiter-controller-update
index a7d36f4..246787b 100755
--- a/usr/bin/jupiter-controller-update
+++ b/usr/bin/jupiter-controller-update
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -euo pipefail
 
diff --git a/usr/bin/jupiter-initial-firmware-update b/usr/bin/jupiter-initial-firmware-update
index 53f335f..db4f759 100755
--- a/usr/bin/jupiter-initial-firmware-update
+++ b/usr/bin/jupiter-initial-firmware-update
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -euo pipefail
 
diff --git a/usr/bin/steamos-polkit-helpers/jupiter-amp-control b/usr/bin/steamos-polkit-helpers/jupiter-amp-control
index 5f1d205..69dafad 100755
--- a/usr/bin/steamos-polkit-helpers/jupiter-amp-control
+++ b/usr/bin/steamos-polkit-helpers/jupiter-amp-control
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/jupiter-biosupdate b/usr/bin/steamos-polkit-helpers/jupiter-biosupdate
index 1d7b792..d90a443 100755
--- a/usr/bin/steamos-polkit-helpers/jupiter-biosupdate
+++ b/usr/bin/steamos-polkit-helpers/jupiter-biosupdate
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/jupiter-check-support b/usr/bin/steamos-polkit-helpers/jupiter-check-support
index 9a442c8..4df43e8 100755
--- a/usr/bin/steamos-polkit-helpers/jupiter-check-support
+++ b/usr/bin/steamos-polkit-helpers/jupiter-check-support
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/jupiter-dock-updater b/usr/bin/steamos-polkit-helpers/jupiter-dock-updater
index 15c695b..ef206ba 100755
--- a/usr/bin/steamos-polkit-helpers/jupiter-dock-updater
+++ b/usr/bin/steamos-polkit-helpers/jupiter-dock-updater
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/jupiter-fan-control b/usr/bin/steamos-polkit-helpers/jupiter-fan-control
index 2dfa176..975f552 100755
--- a/usr/bin/steamos-polkit-helpers/jupiter-fan-control
+++ b/usr/bin/steamos-polkit-helpers/jupiter-fan-control
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/jupiter-get-als-gain b/usr/bin/steamos-polkit-helpers/jupiter-get-als-gain
index 0337e69..add9e8e 100755
--- a/usr/bin/steamos-polkit-helpers/jupiter-get-als-gain
+++ b/usr/bin/steamos-polkit-helpers/jupiter-get-als-gain
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-devkit-mode b/usr/bin/steamos-polkit-helpers/steamos-devkit-mode
index b93f523..70ca709 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-devkit-mode
+++ b/usr/bin/steamos-polkit-helpers/steamos-devkit-mode
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-disable-wireless-power-management b/usr/bin/steamos-polkit-helpers/steamos-disable-wireless-power-management
index 82b602c..4d0c684 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-disable-wireless-power-management
+++ b/usr/bin/steamos-polkit-helpers/steamos-disable-wireless-power-management
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-enable-sshd b/usr/bin/steamos-polkit-helpers/steamos-enable-sshd
index fa7698f..2f129a8 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-enable-sshd
+++ b/usr/bin/steamos-polkit-helpers/steamos-enable-sshd
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-factory-reset-config b/usr/bin/steamos-polkit-helpers/steamos-factory-reset-config
index 3a2a367..580beb0 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-factory-reset-config
+++ b/usr/bin/steamos-polkit-helpers/steamos-factory-reset-config
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-format-device b/usr/bin/steamos-polkit-helpers/steamos-format-device
index d9866a9..a133624 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-format-device
+++ b/usr/bin/steamos-polkit-helpers/steamos-format-device
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-format-sdcard b/usr/bin/steamos-polkit-helpers/steamos-format-sdcard
index 6d9859f..e43e952 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-format-sdcard
+++ b/usr/bin/steamos-polkit-helpers/steamos-format-sdcard
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-poweroff-now b/usr/bin/steamos-polkit-helpers/steamos-poweroff-now
index 1841ce0..9cfa9a6 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-poweroff-now
+++ b/usr/bin/steamos-polkit-helpers/steamos-poweroff-now
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-priv-write b/usr/bin/steamos-polkit-helpers/steamos-priv-write
index d1456c8..8fcd1d4 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-priv-write
+++ b/usr/bin/steamos-polkit-helpers/steamos-priv-write
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -euo pipefail
 shopt -s nullglob
diff --git a/usr/bin/steamos-polkit-helpers/steamos-reboot-now b/usr/bin/steamos-polkit-helpers/steamos-reboot-now
index b4dcc81..5ff87b5 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-reboot-now
+++ b/usr/bin/steamos-polkit-helpers/steamos-reboot-now
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-reboot-other b/usr/bin/steamos-polkit-helpers/steamos-reboot-other
index ad466f2..073927b 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-reboot-other
+++ b/usr/bin/steamos-polkit-helpers/steamos-reboot-other
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-restart-sddm b/usr/bin/steamos-polkit-helpers/steamos-restart-sddm
index 9e08cd6..0711514 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-restart-sddm
+++ b/usr/bin/steamos-polkit-helpers/steamos-restart-sddm
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-select-branch b/usr/bin/steamos-polkit-helpers/steamos-select-branch
index 1b1d6c1..73c1bb2 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-select-branch
+++ b/usr/bin/steamos-polkit-helpers/steamos-select-branch
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-set-hostname b/usr/bin/steamos-polkit-helpers/steamos-set-hostname
index f63e164..56c7905 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-set-hostname
+++ b/usr/bin/steamos-polkit-helpers/steamos-set-hostname
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-set-timezone b/usr/bin/steamos-polkit-helpers/steamos-set-timezone
index 2811d82..427314a 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-set-timezone
+++ b/usr/bin/steamos-polkit-helpers/steamos-set-timezone
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-trim-devices b/usr/bin/steamos-polkit-helpers/steamos-trim-devices
index 38f92a1..296d5d6 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-trim-devices
+++ b/usr/bin/steamos-polkit-helpers/steamos-trim-devices
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/bin/steamos-polkit-helpers/steamos-update b/usr/bin/steamos-polkit-helpers/steamos-update
index bae7403..8a69dd8 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-update
+++ b/usr/bin/steamos-polkit-helpers/steamos-update
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/usr/lib/hwsupport/block-device-event.sh b/usr/lib/hwsupport/block-device-event.sh
index ee67e29..6ff1eb4 100755
--- a/usr/lib/hwsupport/block-device-event.sh
+++ b/usr/lib/hwsupport/block-device-event.sh
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -euo pipefail
 
diff --git a/usr/lib/hwsupport/format-device.sh b/usr/lib/hwsupport/format-device.sh
index 65d77f8..315c588 100755
--- a/usr/lib/hwsupport/format-device.sh
+++ b/usr/lib/hwsupport/format-device.sh
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -e
 
diff --git a/usr/lib/hwsupport/format-sdcard.sh b/usr/lib/hwsupport/format-sdcard.sh
index 36df682..99ed89a 100755
--- a/usr/lib/hwsupport/format-sdcard.sh
+++ b/usr/lib/hwsupport/format-sdcard.sh
@@ -1,3 +1,3 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 exec /usr/lib/hwsupport/format-device.sh --device /dev/mmcblk0 "$@"
diff --git a/usr/lib/hwsupport/steamos-automount.sh b/usr/lib/hwsupport/steamos-automount.sh
index 386f4d8..60bbc93 100755
--- a/usr/lib/hwsupport/steamos-automount.sh
+++ b/usr/lib/hwsupport/steamos-automount.sh
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -euo pipefail
 
diff --git a/usr/lib/hwsupport/trim-devices.sh b/usr/lib/hwsupport/trim-devices.sh
index f5b3e8e..84897df 100755
--- a/usr/lib/hwsupport/trim-devices.sh
+++ b/usr/lib/hwsupport/trim-devices.sh
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -e
 
diff --git a/usr/lib/udev/rules.d/99-sdcard-rescan.rules b/usr/lib/udev/rules.d/99-sdcard-rescan.rules
index 05b6167..f184d9e 100644
--- a/usr/lib/udev/rules.d/99-sdcard-rescan.rules
+++ b/usr/lib/udev/rules.d/99-sdcard-rescan.rules
@@ -1,3 +1,3 @@
 # Work around the misdetection of some SanDisk microSD cards
 
-ACTION=="add", SUBSYSTEM=="mmc", ENV{MMC_NAME}=="SD032", RUN+="/bin/systemd-run --no-block --collect /usr/lib/hwsupport/sdcard-rescan.sh %k"
+ACTION=="add", SUBSYSTEM=="mmc", ENV{MMC_NAME}=="SD032", RUN+="/usr/bin/systemd-run --no-block --collect /usr/lib/hwsupport/sdcard-rescan.sh %k"
diff --git a/usr/lib/udev/rules.d/99-steamos-automount.rules b/usr/lib/udev/rules.d/99-steamos-automount.rules
index 49a5968..18b5149 100644
--- a/usr/lib/udev/rules.d/99-steamos-automount.rules
+++ b/usr/lib/udev/rules.d/99-steamos-automount.rules
@@ -5,7 +5,7 @@ KERNEL!="mmcblk*|sd*",          GOTO="steamos_automount_end"
 ENV{ID_FS_USAGE}!="filesystem", GOTO="steamos_automount_end"
 
 # The service expects to be asynchronous and shouldn't block udev rules
-ACTION=="add",    RUN+="/bin/systemd-run --no-block --collect /usr/lib/hwsupport/block-device-event.sh add %k"
-ACTION=="remove", RUN+="/bin/systemd-run --no-block --collect /usr/lib/hwsupport/block-device-event.sh remove %k"
+ACTION=="add",    RUN+="/usr/bin/systemd-run --no-block --collect /usr/lib/hwsupport/block-device-event.sh add %k"
+ACTION=="remove", RUN+="/usr/bin/systemd-run --no-block --collect /usr/lib/hwsupport/block-device-event.sh remove %k"
 
 LABEL="steamos_automount_end"
diff --git a/usr/share/jupiter_bios_updater/H2OFFTx64-G.sh b/usr/share/jupiter_bios_updater/H2OFFTx64-G.sh
index 8089778..a47f449 100755
--- a/usr/share/jupiter_bios_updater/H2OFFTx64-G.sh
+++ b/usr/share/jupiter_bios_updater/H2OFFTx64-G.sh
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 TARGET_DIR=.
 EXE_FILE=h2offt-g
diff --git a/usr/share/jupiter_bios_updater/H2OFFTx64.sh b/usr/share/jupiter_bios_updater/H2OFFTx64.sh
index ab3513f..6259a69 100755
--- a/usr/share/jupiter_bios_updater/H2OFFTx64.sh
+++ b/usr/share/jupiter_bios_updater/H2OFFTx64.sh
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 TARGET_DIR=.
 EXE_FILE=h2offt
diff --git a/usr/share/jupiter_controller_fw_updater/d20bootloader.py b/usr/share/jupiter_controller_fw_updater/d20bootloader.py
index 6b0ba2b..32c0329 100755
--- a/usr/share/jupiter_controller_fw_updater/d20bootloader.py
+++ b/usr/share/jupiter_controller_fw_updater/d20bootloader.py
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/python3
 # -*- coding: utf-8 -*-
 import click
 import crcmod
diff --git a/usr/share/jupiter_controller_fw_updater/d21bootloader16.py b/usr/share/jupiter_controller_fw_updater/d21bootloader16.py
index 5ee6d41..c662149 100755
--- a/usr/share/jupiter_controller_fw_updater/d21bootloader16.py
+++ b/usr/share/jupiter_controller_fw_updater/d21bootloader16.py
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/python3
 import click
 import crcmod
 import datetime
