diff --git a/usr/lib/hwsupport/steamos-automount.sh b/usr/lib/hwsupport/steamos-automount.sh
index bf58da0..390fa6d 100755
--- a/usr/lib/hwsupport/steamos-automount.sh
+++ b/usr/lib/hwsupport/steamos-automount.sh
@@ -2,6 +2,9 @@
 
 set -euo pipefail
 
+IMAGE_INFO="/usr/share/ublue-os/image-info.json"
+BASE_IMAGE_NAME=$(jq -r '."base-image-name"' < $IMAGE_INFO)
+
 # Originally from https://serverfault.com/a/767079
 
 # This script is called from our systemd unit file to mount or unmount
@@ -198,6 +201,10 @@ do_mount()
             ;;
     esac
 
+    if [[ $BASE_IMAGE_NAME =~ "silverblue" ]]; then
+        touch "${mount_point}/.trackerignore"
+    fi
+
     #### SteamOS Btrfs Begin ####
     if [[ "${ID_FS_TYPE}" == "btrfs" ]]; then
         # Workaround for for Steam compression bug
