diff --git a/usr/lib/hwsupport/block-device-event.sh b/usr/lib/hwsupport/block-device-event.sh
index 132d94c..f3606b5 100755
--- a/usr/lib/hwsupport/block-device-event.sh
+++ b/usr/lib/hwsupport/block-device-event.sh
@@ -40,7 +40,7 @@ do_add()
     detected_us=$(make_dbus_udisks_call get-property data t "${drive}" Drive TimeMediaDetected)
     # The 5 seconds window is taken from the original GNOME fix that inspired this one
     # https://gitlab.gnome.org/GNOME/gvfs/-/commit/b4800b987b4a8423a52306c9aef35b3777464cc5
-    if (( detected_us / 1000000 + 5 < current_time )); then
+    if (( detected_us / 1000000 + 30 < current_time )); then
         echo "Skipping mounting /dev/${DEVBASE} because it has not been inserted recently" >&2
         exit 0
     fi
