diff --git a/etc/systemd/system/steamos-automount@.service b/etc/systemd/system/steamos-automount@.service
index 96e8818..a16650c 100644
--- a/etc/systemd/system/steamos-automount@.service
+++ b/etc/systemd/system/steamos-automount@.service
@@ -14,5 +14,5 @@ Description=Mount removable drive %i
 [Service]
 Type=oneshot
 RemainAfterExit=true
-ExecStart=/usr/lib/hwsupport/steamos-automount.sh add %i
-ExecStop=/usr/lib/hwsupport/steamos-automount.sh remove %i
+ExecStart=/usr/libexec/steamos-automount add %i
+ExecStop=/usr/libexec/steamos-automount remove %i

diff --git a/usr/bin/steamos-polkit-helpers/steamos-format-sdcard b/usr/bin/steamos-polkit-helpers/steamos-format-sdcard
index 53b4f2a..9022762 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-format-sdcard
+++ b/usr/bin/steamos-polkit-helpers/steamos-format-sdcard
@@ -8,4 +8,4 @@ then
     exec pkexec --disable-internal-agent "$0" --owner "$UID":"$gid" "$@"
 fi
 
-exec /usr/lib/hwsupport/format-sdcard.sh "$@"
+exec /usr/libexec/format-sdcard "$@"
diff --git a/usr/bin/steamos-polkit-helpers/steamos-trim-devices b/usr/bin/steamos-polkit-helpers/steamos-trim-devices
index 296a6c7..698ee59 100755
--- a/usr/bin/steamos-polkit-helpers/steamos-trim-devices
+++ b/usr/bin/steamos-polkit-helpers/steamos-trim-devices
@@ -7,4 +7,4 @@ then
     exec pkexec --disable-internal-agent "$0" "$@"
 fi
 
-exec /usr/lib/hwsupport/trim-devices.sh "$@"
+exec /usr/libexec/trim-devices "$@"
diff --git a/usr/lib/hwsupport/format-sdcard.sh b/usr/lib/hwsupport/format-sdcard.sh
index 3e13083..39f08ac 100755
--- a/usr/lib/hwsupport/format-sdcard.sh
+++ b/usr/lib/hwsupport/format-sdcard.sh
@@ -1,3 +1,3 @@
 #!/usr/bin/sh
 
-exec /usr/lib/hwsupport/format-device.sh --device /dev/mmcblk0 "$@"
+exec /usr/libexec/format-device --device /dev/mmcblk0 "$@"
