--- a/galileo-mura-setup
+++ b/galileo-mura-setup
@@ -16,7 +16,7 @@
 deck_serial=${deck_serial^^}
 manufacturer=$(sed -n 's/.*Manufacturer: \(.*\)/\1/p' < /tmp/mura/info.txt)
 
-mura_base_path=/home/deck/.config/gamescope/mura
+mura_base_path="${XDG_CONFIG_HOME:-$HOME/.config}/gamescope/mura"
 mkdir -p "$mura_base_path"
 
 if [ "$manufacturer" != "SDC" ]; then
@@ -74,7 +74,7 @@
 fi
 
 # Grody! No gamescopectl yet kskdsjsadjjksjkjsj
-xprop -root -f GAMESCOPE_COLOR_MURA_CORRECTION_IMAGE 8s -set GAMESCOPE_COLOR_MURA_CORRECTION_IMAGE "$mura_path/$display_serial"
+DISPLAY=:0 xprop -root -f GAMESCOPE_COLOR_MURA_CORRECTION_IMAGE 8s -set GAMESCOPE_COLOR_MURA_CORRECTION_IMAGE "$mura_path/$display_serial"
 
 echo "[galileo-mura] Success! All good. Using mura at: $mura_path"
 
