--- a/etc/xdg/kdeglobals
+++ b/etc/xdg/kdeglobals
@@ -10,8 +10,13 @@
 [KDE Control Module Restrictions][$i]
 kcm_sddm.desktop=false
 kcm_plymouth.desktop=false
+kcm_updates=false
+
+[KScreen]
+XwaylandClientsScale=false
 
 [General]
+fixed=Fira Code,9,-1,5,50,0,0,0,0,0
 font=Noto Sans,11,-1,5,50,0,0,0,0,0
 menuFont=Noto Sans,11,-1,5,50,0,0,0,0,0
 smallestReadableFont=Noto Sans,9,-1,5,50,0,0,0,0,0
