diff --git a/usr/bin/steamos-nested-desktop b/usr/bin/steamos-nested-desktop
index 7cc1b05..22b345f 100755
--- a/usr/bin/steamos-nested-desktop
+++ b/usr/bin/steamos-nested-desktop
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/usr/bin/bash
 
 # Remove the performance overlay, it meddles with some tasks
 unset LD_PRELOAD
@@ -7,8 +7,10 @@ unset LD_PRELOAD
 ## whilst being launched by plasma-session
 mkdir $XDG_RUNTIME_DIR/nested_plasma -p
 cat <<EOF > $XDG_RUNTIME_DIR/nested_plasma/kwin_wayland_wrapper
-#!/bin/sh
-/usr/bin/kwin_wayland_wrapper --width 1280 --height 800 --no-lockscreen \$@
+#!/usr/bin/bash
+source /etc/default/steamos-nested-desktop
+
+/usr/bin/kwin_wayland_wrapper --width ${STEAMOS_NESTED_DESKTOP_WIDTH:-1280} --height ${STEAMOS_NESTED_DESKTOP_HEIGHT:-800} --no-lockscreen \$@
 EOF
 chmod a+x $XDG_RUNTIME_DIR/nested_plasma/kwin_wayland_wrapper
 export PATH=$XDG_RUNTIME_DIR/nested_plasma:$PATH
