diff --git a/pipewire-confs/hardware-profiles/pipewire-hwconfig b/pipewire-confs/hardware-profiles/pipewire-hwconfig
index 11c8bd2..ac7adb6 100755
--- a/pipewire-confs/hardware-profiles/pipewire-hwconfig
+++ b/pipewire-confs/hardware-profiles/pipewire-hwconfig
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/pipewire-confs/hardware-profiles/valve-galileo/pipewire.conf.d/filter-chain.conf b/pipewire-confs/hardware-profiles/valve-galileo/pipewire.conf.d/filter-chain.conf
index 262fa0f..d7055e2 100644
--- a/pipewire-confs/hardware-profiles/valve-galileo/pipewire.conf.d/filter-chain.conf
+++ b/pipewire-confs/hardware-profiles/valve-galileo/pipewire.conf.d/filter-chain.conf
@@ -14,7 +14,7 @@ context.modules = [
                     {
                         type = ladspa
                         name = rnnoise
-                        plugin = /usr/lib/ladspa/librnnoise_ladspa.so
+                        plugin = /usr/lib64/ladspa/librnnoise_ladspa.so
                         label = noise_suppressor_stereo
                         control = {
                             "VAD Threshold (%)" = 23.0
diff --git a/pipewire-confs/hardware-profiles/valve-jupiter/pipewire.conf.d/filter-chain.conf b/pipewire-confs/hardware-profiles/valve-jupiter/pipewire.conf.d/filter-chain.conf
index 90a1e12..43f03e0 100644
--- a/pipewire-confs/hardware-profiles/valve-jupiter/pipewire.conf.d/filter-chain.conf
+++ b/pipewire-confs/hardware-profiles/valve-jupiter/pipewire.conf.d/filter-chain.conf
@@ -14,7 +14,7 @@ context.modules = [
                     {
                         type = ladspa
                         name = rnnoise
-                        plugin = /usr/lib/ladspa/librnnoise_ladspa.so
+                        plugin = /usr/lib64/ladspa/librnnoise_ladspa.so
                         label = noise_suppressor_stereo
                         control = {
                             "VAD Threshold (%)" = 23.0
diff --git a/pipewire-confs/systemd/system/pipewire-sysconf.service b/pipewire-confs/systemd/system/pipewire-sysconf.service
index ffbdd97..449a278 100644
--- a/pipewire-confs/systemd/system/pipewire-sysconf.service
+++ b/pipewire-confs/systemd/system/pipewire-sysconf.service
@@ -17,7 +17,7 @@ Before=shutdown.target
 [Service]
 Type=oneshot
 RemainAfterExit=yes
-ExecStart=/usr/share/pipewire/hardware-profiles/pipewire-hwconfig
+ExecStart=/usr/libexec/hwsupport/pipewire-hwconfig
 
 [Install]
 WantedBy=multi-user.target
diff --git a/scripts/install_plugins.sh b/scripts/install_plugins.sh
index ad10f23..320aa0a 100755
--- a/scripts/install_plugins.sh
+++ b/scripts/install_plugins.sh
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/usr/bin/bash
 
 set -eu
 
@@ -9,9 +9,9 @@ if [ -e "$HARDCODED_PLUGIN_DEST_1" ]; then rm -r "$HARDCODED_PLUGIN_DEST_1"; fi;
 if [ -e "$HARDCODED_PLUGIN_DEST_2" ]; then rm -r "$HARDCODED_PLUGIN_DEST_2"; fi;
 
 
-mkdir -p "$DEST_DIR"/usr/lib/lv2
-cp -r bin/lv2/g++/valve_deck_speakers.lv2 "$DEST_DIR"/usr/lib/lv2/valve_deck_speakers.lv2
-cp -r bin/lv2/g++/valve_binaural.lv2 "$DEST_DIR"/usr/lib/lv2/valve_binaural.lv2
-cp -r bin/lv2/g++/valve_deck_microphone.lv2 "$DEST_DIR"/usr/lib/lv2/valve_deck_microphone.dsp
-# cp -r bin/svg $DEST_DIR/usr/lib/lv2/svg
-# cp version.txt $DEST_DIR/usr/lib/lv2/valve_audio_dsp_version.txt
+mkdir -p "$LIB_DIR"/lv2
+cp -r bin/lv2/g++/valve_deck_speakers.lv2 "$LIB_DIR"/lv2/valve_deck_speakers.lv2
+cp -r bin/lv2/g++/valve_binaural.lv2 "$LIB_DIR"/lv2/valve_binaural.lv2
+cp -r bin/lv2/g++/valve_deck_microphone.lv2 "$LIB_DIR"/lv2/valve_deck_microphone.dsp
+# cp -r bin/svg $LIB_DIR/lv2/svg
+# cp version.txt $LIB_DIR/lv2/valve_audio_dsp_version.txt
diff --git a/wireplumber/hardware-profiles/wireplumber-hwconfig b/wireplumber/hardware-profiles/wireplumber-hwconfig
index 842edb2..3f03354 100755
--- a/wireplumber/hardware-profiles/wireplumber-hwconfig
+++ b/wireplumber/hardware-profiles/wireplumber-hwconfig
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
diff --git a/wireplumber/systemd/system/wireplumber-sysconf.service b/wireplumber/systemd/system/wireplumber-sysconf.service
index 92b811c..cdf5ed9 100644
--- a/wireplumber/systemd/system/wireplumber-sysconf.service
+++ b/wireplumber/systemd/system/wireplumber-sysconf.service
@@ -17,7 +17,7 @@ Before=shutdown.target
 [Service]
 Type=oneshot
 RemainAfterExit=yes
-ExecStart=/usr/share/wireplumber/hardware-profiles/wireplumber-hwconfig
+ExecStart=/usr/libexec/hwsupport/wireplumber-hwconfig
 
 [Install]
 WantedBy=multi-user.target
