diff --git a/src/steamcompmgr.cpp b/src/steamcompmgr.cpp
index 05337c1..aeb3a39 100644
--- a/src/steamcompmgr.cpp
+++ b/src/steamcompmgr.cpp
@@ -4949,8 +4949,10 @@ handle_property_notify(xwayland_ctx_t *ctx, XPropertyEvent *ev)
 			focusDirty = true;
 		}
 	}
-	if (ev->atom == ctx->atoms.steamTouchClickModeAtom )
+	if (ev->atom == ctx->atoms.steamTouchClickModeAtom && !env_to_bool(getenv("GAMESCOPE_DISABLE_TOUCH_CLICK_MODE")))
 	{
+		// When steam changes this to 1 stuff breaks, make it not shoot itself
+		// in the foot
 		gamescope::cv_touch_click_mode = (gamescope::TouchClickMode) get_prop(ctx, ctx->root, ctx->atoms.steamTouchClickModeAtom, 0u );
 	}
 	if (ev->atom == ctx->atoms.steamStreamingClientAtom)
