diff --git a/src/shaders/cs_composite_blit.comp b/src/shaders/cs_composite_blit.comp
index 52bd9d517..359ee3a19 100644
--- a/src/shaders/cs_composite_blit.comp
+++ b/src/shaders/cs_composite_blit.comp
@@ -50,7 +50,7 @@ void main() {
     }
 
     outputValue.rgb = encodeOutputColor(outputValue.rgb);
-    imageStore(dst, ivec2(coord), outputValue);
+    imageStore(dst, ivec2(coord), vec4(outputValue.rgb, 1));
 
     // Indicator to quickly tell if we're in the compositing path or not.
     if (checkDebugFlag(compositedebug_Markers))
