diff --git a/Makefile b/Makefile
index 71e2122..e8bf427 100644
--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,5 @@
 PREFIX ?= $(HOME)/.local
+INSTALL_PREFIX ?= $(PREFIX)
 
 ##@ General
 
@@ -15,7 +16,7 @@ install: ## Install OpenGamepadUI (default: ~/.local)
 	mkdir -p $(PREFIX)/share/applications
 	install -Dm644 usr/share/applications/opengamepadui.desktop \
 		$(PREFIX)/share/applications/opengamepadui.desktop
-	sed -i 's|Exec=opengamepadui|Exec=$(PREFIX)/bin/opengamepadui|g' \
+	sed -i 's|Exec=opengamepadui|Exec=$(INSTALL_PREFIX)/bin/opengamepadui|g' \
 		$(PREFIX)/share/applications/opengamepadui.desktop
 	mkdir -p $(PREFIX)/share/icons/hicolor/scalable/apps
 	install -Dm644 usr/share/icons/hicolor/scalable/apps/opengamepadui.svg \
diff --git a/usr/bin/opengamepadui b/usr/bin/opengamepadui
index e5ebb43..5395eb7 100755
--- a/usr/bin/opengamepadui
+++ b/usr/bin/opengamepadui
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
 PREFIX=$(dirname -- "${SCRIPT_DIR}")
 
diff --git a/usr/share/opengamepadui/scripts/powertools b/usr/share/opengamepadui/scripts/powertools
index 79d6740..9ab3ddf 100755
--- a/usr/share/opengamepadui/scripts/powertools
+++ b/usr/share/opengamepadui/scripts/powertools
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/bash
 
 set -eu
 
