diff --git a/src/main.conf b/src/main.conf
index d108934a8..ff0e98434 100644
--- a/src/main.conf
+++ b/src/main.conf
@@ -60,14 +60,14 @@
 # or both Multiple Profiles Single Device (MPSD) and Multiple Profiles Multiple
 # Devices (MPMD) configurations.
 # Possible values: "off", "single", "multiple"
-#MultiProfile = off
+MultiProfile = multiple
 
 # Permanently enables the Fast Connectable setting for adapters that
 # support it. When enabled other devices can connect faster to us,
 # however the tradeoff is increased power consumptions. This feature
 # will fully work only on kernel version 4.1 and newer. Defaults to
 # 'false'.
-#FastConnectable = false
+FastConnectable = true
 
 # Default privacy setting.
 # Enables use of private address.
@@ -193,8 +193,8 @@
 
 # LE scanning parameters used for passive scanning supporting wake from suspend
 # scenarios
-#ScanIntervalSuspend=
-#ScanWindowSuspend=
+ScanIntervalSuspend=2240
+ScanWindowSuspend=224
 
 # LE scanning parameters used for active scanning supporting discovery
 # proceedure
