diff --git a/src/main.conf b/src/main.conf
index 5b2dc83..a24fb0b 100644
--- a/src/main.conf
+++ b/src/main.conf
@@ -141,7 +141,7 @@
 # a6695ace-ee7f-4fb9-881a-5fac66c629af (BlueZ Experimental Offload Codecs)
 # 6fbaf188-05e0-496a-9885-d6ddfdb4e03e (BlueZ Experimental ISO socket)
 # Defaults to false.
-#KernelExperimental = false
+KernelExperimental = 15c0a148-c273-11ea-b3de-0242ac130004
 
 # The duration to avoid retrying to resolve a peer's name, if the previous
 # try failed.
