diff -ru bluez-5.66.orig/profiles/audio/avrcp.c bluez-5.66/profiles/audio/avrcp.c
--- bluez-5.66.orig/profiles/audio/avrcp.c	2022-03-16 08:06:20.000000000 -0700
+++ bluez-5.66/profiles/audio/avrcp.c	2023-07-30 08:03:17.414213611 -0700
@@ -1210,6 +1210,10 @@
 					GUINT_TO_POINTER(str_to_metadata(key)));
 	}
 
+	if (attrs == NULL)
+		return g_list_prepend(NULL,
+				GUINT_TO_POINTER(AVRCP_MEDIA_ATTRIBUTE_TITLE));
+
 	return attrs;
 }
 
Only in bluez-5.66/profiles/audio: avrcp.c.orig
