diff --git a/vite.config.js b/vite.config.js
index 568aae41..d24f1c29 100644
--- a/vite.config.js
+++ b/vite.config.js
@@ -53,13 +53,6 @@ export default defineConfig({
     plugins: [
         vue(),
         ViteEjsPlugin({ header }),
-        // The Codecov vite plugin should be after all other plugins
-        codecovVitePlugin({
-            enableBundleAnalysis: true,
-            bundleName: "sunshine",
-            uploadToken: process.env.CODECOV_TOKEN,
-            gitService: "github",
-        }),
     ],
     root: resolve(assetsSrcPath),
     build: {
