diff --git a/src/util/00-mesa-defaults.conf b/src/util/00-mesa-defaults.conf
index 603a6dd..0cf8099 100644
--- a/src/util/00-mesa-defaults.conf
+++ b/src/util/00-mesa-defaults.conf
@@ -729,6 +729,7 @@ TODO: document the other workarounds.
         <application name="gnome-shell" executable="gnome-shell">
             <option name="adaptive_sync" value="false" />
             <option name="v3d_nonmsaa_texture_size_limit" value="true" />
+            <option name="mesa_glthread" value="false"/>
         </application>
         <application name="Desktop — Plasma" executable="plasmashell">
             <option name="adaptive_sync" value="false" />
