From fcf83b479620d5bea6b7c81bc1255ecc5b12daac Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Robert-Andr=C3=A9=20Mauchin?= <zebob.m@gmail.com>
Date: Tue, 21 Jun 2022 17:28:40 +0200
Subject: [PATCH] Add RPM_BUILD_ROOT to env

---
 SConstruct | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/SConstruct b/SConstruct
index 7e12d413..4a4f1497 100755
--- a/SConstruct
+++ b/SConstruct
@@ -538,7 +538,7 @@ options = dict(
     PREFIX=GetOption('prefix'),
     ENV = dict([ (key, os.environ[key])
                  for key in os.environ
-                 if key in ['PATH', 'TERM', 'HOME', 'PKG_CONFIG_PATH']
+                 if key in ['PATH', 'TERM', 'HOME', 'PKG_CONFIG_PATH', 'RPM_BUILD_ROOT']
               ])
 )
 
-- 
2.36.1

