--- iwd-2.13/src/iwd-dbus.conf	2023-11-19 18:55:17.000000000 +0100
+++ iwd-2.13/src/iwd-dbus.conf.new	2024-01-16 22:13:06.696991823 +0100
@@ -19,7 +19,7 @@
     <allow send_destination="net.connman.iwd"/>
   </policy>
 
-  <policy group="netdev">
+  <policy group="network">
     <allow send_destination="net.connman.iwd"/>
   </policy>
 
--- iwd-2.13/wired/ead-dbus.conf	2024-01-22 18:16:13.000000000 +0100
+++ iwd-2.13/wired/ead-dbus.conf.new	2024-01-25 16:21:33.383168588 +0100
@@ -17,7 +17,7 @@
     <allow send_destination="net.connman.ead"/>
   </policy>
 
-  <policy group="netdev">
+  <policy group="network">
     <allow send_destination="net.connman.ead"/>
   </policy>
 
