From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Antheas Kapenekakis <git@antheas.dev>
Date: Sun, 27 Apr 2025 21:50:17 +0200
Subject: rename back to --json-fd

Signed-off-by: Antheas Kapenekakis <git@antheas.dev>
---
 crates/lib/src/cli.rs | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/crates/lib/src/cli.rs b/crates/lib/src/cli.rs
index 3412a6b03c48..5cda8810781f 100644
--- a/crates/lib/src/cli.rs
+++ b/crates/lib/src/cli.rs
@@ -51,16 +51,19 @@ pub(crate) struct ProgressOptions {
     /// File descriptor number which must refer to an open pipe.
     ///
     /// Progress is written as JSON lines to this file descriptor.
-    #[clap(long, hide = true)]
+    #[clap(long)]
     pub(crate) progress_fd: Option<RawProgressFd>,
+
+    #[clap(long)]
+    pub(crate) json_fd: Option<RawProgressFd>,
 }
 
 impl TryFrom<ProgressOptions> for ProgressWriter {
     type Error = anyhow::Error;
 
     fn try_from(value: ProgressOptions) -> Result<Self> {
-        let r = value
-            .progress_fd
+        let fd = value.progress_fd.or(value.json_fd);
+        let r = fd
             .map(TryInto::try_into)
             .transpose()?
             .unwrap_or_default();
-- 
2.51.0

