diff --git a/tools/helpers/mount.py b/tools/helpers/mount.py
index a0d34be5..f507d668 100644
--- a/tools/helpers/mount.py
+++ b/tools/helpers/mount.py
@@ -133,7 +133,7 @@ def mount(args, source, destination, create_folders=True, umount=False,
                             destination)
 
     extra_args = []
-    opt_args = []
+    opt_args = ["context=\"system_u:object_r:waydroid_rootfs_t:s0\""]
     if mount_type:
         extra_args.extend(["-t", mount_type])
     if readonly:
