From 66c8343c4d2ea118601ba5d8ce52fa622cbcd665 Mon Sep 17 00:00:00 2001
From: Aarron Lee <aclee91@gmail.com>
Date: Sun, 19 Nov 2023 16:11:44 -0500
Subject: [PATCH] update regex for deprecation warning

---
 tools/helpers/net.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tools/helpers/net.py b/tools/helpers/net.py
index c20a95bc..b267357a 100644
--- a/tools/helpers/net.py
+++ b/tools/helpers/net.py
@@ -31,6 +31,6 @@ def get_device_ip_address():
 
     try:
         with open(lease_file) as f:
-            return re.search("(\d{1,3}\.){3}\d{1,3}\s", f.read()).group().strip()
+            return re.search(r"(\d{1,3}\.){3}\d{1,3}\s", f.read()).group().strip()
     except:
-        pass
\ No newline at end of file
+        pass
