--- waydroid-1.2.0/data/scripts/waydroid-net.sh	2021-10-19 15:50:56.000000000 +0800
+++ waydroid-1.2.0-1/data/scripts/waydroid-net.sh	2021-10-21 22:57:50.991706580 +0800
@@ -1,4 +1,4 @@
-#!/bin/sh -
+#!/usr/bin/sh -
 
 varrun="/run/waydroid-lxc"
 varlib="/var/lib"
@@ -60,6 +60,8 @@ _ifup() {
     ip addr add ${CIDR_ADDR} broadcast + dev ${LXC_BRIDGE}
     ip link set dev ${LXC_BRIDGE} address $LXC_BRIDGE_MAC
     ip link set dev ${LXC_BRIDGE} up
+    # Fedora specific
+    firewall-cmd --zone=trusted --add-interface=${LXC_BRIDGE} || true
 }
 
 start_ipv6() {
